traffic_ops/app/lib/API/Tenant.pm

```diff
@@ -131,21 +130,74 @@ sub update {
 
 	my $is_active = $params->{active};
 	
-	if ( !$params->{active} && $self->isRootTenant($id)) {
-		return $self->alert("Root user cannot be in-active.");
+	if ( !$params->{active} && $tenant_utils->is_root_tenant($tenants_data, $id)) {
+		return $self->alert("Root tenant cannot be in-active.");
+	}
+
+	#this is a write operation, allowed only by parents of the tenant (which are the owners of the resource of type tenant)	
+	my $current_resource_tenancy = $self->db->resultset('Tenant')->search( { id => $id } )->get_column('parent_id')->single();
+	if (!defined($current_resource_tenancy)) {
+		#no parent - the tenant is its-own owner
+		$current_resource_tenancy = $id;
 	}
 	
+	if (!$tenant_utils->is_tenant_resource_accessible($tenants_data, $current_resource_tenancy)) {
+		return $self->forbidden(); #Current owning tenant is not under user's tenancy
+	}
 
-	if ( !defined($params->{parentId}) && !isRootTenant($id) ) {
-		return $self->alert("Only the \"root\" tenant can have no parent.");
+	if (!$tenant_utils->is_tenant_resource_accessible($tenants_data, $params->{parentId})) {
+		return $self->forbidden(); #Parent tenant to be set is not under user's tenancy
 	}
+
+
+	if ($params->{parentId} != $tenant->parent) {
+		#parent replacement
+		if (!defined($tenant_utils->get_tenant_by_id($tenants_data, $params->{parentId}))) {
+			return $self->alert("Parent tenant does not exists.");
+		}
+		my $parent_depth = $tenant_utils->get_tenant_heirarchy_depth($tenants_data, $params->{parentId});
+		if (!defined($parent_depth))
+		{
+			return $self->alert("Failed to retrieve parent tenant depth.");
+		}
+
+		my $tenant_height = $tenant_utils->get_tenant_heirarchy_height($tenants_data, $id);
+		if (!defined($tenant_height))
+		{
+			return $self->alert("Failed to retrieve tenant height.");
+		}
+	
+		if ($parent_depth+$tenant_height+1 > $tenant_utils->max_heirarchy_limit())
```

Comment:
so i've been thinking about this a little more. by putting in a max hierarchy depth, you are putting in a safety net in case somehow cyclical parents are created. for example.

root has no parent
foo has baz parent <-- cyclical
bar has foo parent
baz has bar parent

however, this feels pretty hacky putting in this safety net and it could cover up valid problems. i think it's more important to just ensure that the parent for a tenant can never be assigned as itself or as a child tenant and then make sure you have a robust set of tests to validate this. for example:

- root
-- foo
--- bar
---- baz

when updating the bar tenant, bar and baz are invalid parents.

IF a bug slips thru and an infinite loop is created, then we fix it and write a test case to represent the scenario that created an infinite loop to ensure it is fixed and doesn't happen again. safety nets worry me because they may cover up bugs that need to be fixed.